lib/features/push_notification/presentation/controller/web_socket_controller.dart

```diff
@@ -0,0 +1,84 @@
+import 'package:core/presentation/state/failure.dart';
+import 'package:core/presentation/state/success.dart';
+import 'package:core/utils/app_logger.dart';
+import 'package:jmap_dart_client/jmap/account_id.dart';
+import 'package:jmap_dart_client/jmap/core/session/session.dart';
+import 'package:tmail_ui_user/features/push_notification/domain/exceptions/web_socket_exceptions.dart';
+import 'package:tmail_ui_user/features/push_notification/domain/state/web_socket_push_state.dart';
+import 'package:tmail_ui_user/features/push_notification/domain/usecases/connect_web_socket_interactor.dart';
+import 'package:tmail_ui_user/features/push_notification/presentation/controller/push_base_controller.dart';
+import 'package:tmail_ui_user/features/push_notification/presentation/extensions/state_change_extension.dart';
+import 'package:tmail_ui_user/main/routes/route_navigation.dart';
+
+class WebSocketController extends PushBaseController {
+  WebSocketController._internal();
+
+  static final WebSocketController _instance = WebSocketController._internal();
+
+  static WebSocketController get instance => _instance;
+
+  ConnectWebSocketInteractor? _connectWebSocketInteractor;
+
+  int _retryRemained = 3;
+
+  @override
+  void handleFailureViewState(Failure failure) {
+    logError('WebSocketController::handleFailureViewState():Failure $failure');
+    cancelStateStreamSubscription();
```

Comment:
It should not be placed here. In case: Success -> Failure -> Success. It may not be synchronized.

Reply:
Failure state should be the end of current web socket connection to retry for a new one